From b5d51fcc109197b58a38a922aeba5c775031fdf4 Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:25:19 -0400
Subject: [md] raid5: Use slow_path to release stripe when mddev->thread is null

Message-id: <1409563584-9000-18-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 89961
O-Subject: [RHEL7.1 PATCH 17/82] raid5: Use slow_path to release stripe when mddev->thread is null
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: majianpeng <majianpeng@gmail.com>

When release_stripe() is called in grow_one_stripe(), the
mddev->thread is null. So it will omit one wakeup this thread to
release stripe.
For this condition, use slow_path to release stripe.

Bug was introduced in 3.12

Cc: stable@vger.kernel.org (3.12+)
Fixes: 773ca82fa1ee58dd1bf88b
Signed-off-by: Jianpeng Ma <majianpeng@gmail.com>
Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit ad4068de49862b083ac2a15bc50689bb30ce3e44)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/raid5.c b/drivers/md/raid5.c
index 3d90106..0c7c311 100644
--- a/drivers/md/raid5.c
+++ b/drivers/md/raid5.c
@@ -326,7 +326,8 @@ static void release_stripe(struct stripe_head *sh)
  unsigned long flags;
  bool wakeup;
 
- if (test_and_set_bit(STRIPE_ON_RELEASE_LIST, &sh->state))
+ if (unlikely(!conf->mddev->thread) ||
+  test_and_set_bit(STRIPE_ON_RELEASE_LIST, &sh->state))
   goto slow_path;
  wakeup = llist_add(&sh->release_list, &conf->released_stripes);
  if (wakeup)
-- 
1.7.1